File: src/main/java/xyz/gnarbot/gnar/commands/executors/general/BotInfoCommand.java
package xyz.gnarbot.gnar.commands.executors.general;

import net.dv8tion.jda.core.JDA;
import net.dv8tion.jda.core.entities.Guild;
import net.dv8tion.jda.core.entities.Member;
import xyz.gnarbot.gnar.Bot;
import xyz.gnarbot.gnar.commands.handlers.Command;
import xyz.gnarbot.gnar.commands.handlers.CommandExecutor;
import xyz.gnarbot.gnar.commands.handlers.CommandHandler;
import xyz.gnarbot.gnar.servers.Shard;
import xyz.gnarbot.gnar.utils.Note;

import java.util.StringJoiner;
import java.util.stream.Collectors;

@Command(aliases = {"info", "botinfo"}, description = "Show information about GN4R-BOT.")
public class BotInfoCommand extends CommandExecutor
{
    @Override
    public void execute(Note note, String label, String[] args)
    {
        CommandHandler commandHandler = note.getHost().getCommandHandler();
        
        int textChannels = 0;
        int voiceChannels = 0;
        int servers = 0;
        int users = 0;
        int offline = 0;
        int online = 0;
        int inactive = 0;
        int dnd = 0;
        
        for (Shard shard : Bot.INSTANCE.getShards())
        {
            JDA jda = shard.getJDA();
            
            servers += jda.getGuilds().size();
            for (Guild g : jda.getGuilds())
            {
                for (Member u : g.getMembers())
                {
                    switch (u.getOnlineStatus())
                    {
                        case ONLINE:
                            online++;
                            break;
                        case OFFLINE:
                            offline++;
                            break;
                        case IDLE:
                            inactive++;
                            break;
                        case DO_NOT_DISTURB:
                            dnd++;
                            break;
                    }
                }
                users += g.getMembers().size();
                textChannels += g.getTextChannels().size();
                voiceChannels += g.getVoiceChannels().size();
            }
        }
        
        int commandSize = commandHandler.getUniqueRegistry().values().parallelStream()
                .filter(CommandExecutor::isShownInHelp)
                .collect(Collectors.toList())
                .size();
        
        int requests = Bot.INSTANCE.getShards().stream()
                .flatMap(shard -> shard.getHosts().stream())
                .mapToInt(guild -> guild.getCommandHandler().getRequests())
                .sum();
        
        StringJoiner joiner = new StringJoiner("\n");
        
        //         "__**General**                                                      __"
        
        joiner.add("Requests: **[" + requests + "]()**");
        joiner.add("Servers: **[" + servers + "]()**");
        joiner.add("Shards: **[" + Bot.INSTANCE.getShards().size() + "]()**");
        joiner.add("");
        joiner.add("__**Channels**                                                     __");
        joiner.add("  Text: **[" + textChannels + "]()**");
        joiner.add("  Voice: **[" + voiceChannels + "]()**");
        joiner.add("");
        joiner.add("__**Users**                                                          __");
        joiner.add("  Total: **[" + users + "]()**");
        joiner.add("  Online: **[" + online + "]()**");
        joiner.add("  Offline: **[" + offline + "]()**");
        joiner.add("  Inactive: **[" + inactive + "]()**");
        joiner.add("  DND: **[" + dnd + "]()**");
        joiner.add("");
        joiner.add("__**Others**                                                         __");
        joiner.add("  Creators: **[Avarel](https://github.com/Avarel)** and **[Maeyrl](https://github.com/maeyrl)**");
        joiner.add("  Contributor: **[Gatt](https://github.com/RealGatt)**");
        joiner.add("  Website: **[gnarbot.xyz](http://gnarbot.xyz)**");
        joiner.add("  Commands: **[" + commandSize + "]()**");
        joiner.add("  Library: **[JDA 3" + "](https://github.com/DV8FromTheWorld/JDA)**");
        joiner.add("  Uptime: **[" + Bot.INSTANCE.getSimpleUptime() + "]()**");
        
        note.replyEmbedRaw("Bot Information", joiner.toString(), Bot.getColor(), note.getJDA()
                .getSelfUser()
                .getAvatarUrl());
    }
}

File: src/main/kotlin/xyz/gnarbot/gnar/commands/executors/general/BotInfoCommand.kt
package xyz.gnarbot.gnar.commands.executors.general

import link
import net.dv8tion.jda.core.OnlineStatus
import net.dv8tion.jda.core.entities.Message
import xyz.gnarbot.gnar.Constants
import xyz.gnarbot.gnar.commands.Command
import xyz.gnarbot.gnar.commands.CommandExecutor

@Command(
        aliases = arrayOf("info", "botinfo"),
        description = "Show information about the bot."
)
class BotInfoCommand : CommandExecutor() {
    override fun execute(message: Message, args: Array<String>) {
        val registry = bot.commandRegistry

        // Uptime
        val s = bot.uptime / 1000
        val m = s / 60
        val h = m / 60
        val d = h / 24

        var voiceConnections = 0

        var textChannels = 0
        var voiceChannels = 0
        var guildData = 0 // wrapper
        var guilds = 0

        var users = 0
        var offline = 0
        var online = 0
        var inactive = 0

        for (shard in bot.shards) {
            guilds += shard.guilds.size

            for (guild in shard.guilds) {
                for (member in guild.members) {
                    when (member.onlineStatus) {
                        OnlineStatus.ONLINE -> online++
                        OnlineStatus.OFFLINE -> offline++
                        OnlineStatus.IDLE -> inactive++
                        else -> {}
                    }
                }

                if (guild.selfMember.voiceState.channel != null) {
                    voiceConnections++
                }
            }

            users += shard.users.size
            textChannels += shard.textChannels.size
            voiceChannels += shard.voiceChannels.size
            guildData += shard.guildData.size
        }

        val commandSize = registry.entries.count { it.meta.category.show }

        val requests = bot.shards
                .flatMap { it.guildData.values }
                .sumBy { it.commandHandler.requests }

        message.respond().embed("Bot Information") {
            color = Constants.COLOR

            inline {
                field("Requests", requests)
                field("Requests Per Hour", requests / if (h == 0L) 1 else h)
                field("Website", link("gnarbot.xyz", "https://gnarbot.xyz"))

                field("Text Channels", textChannels)
                field("Voice Channels", voiceChannels)
                field("Voice Connections", voiceConnections)

                field("Guilds", guilds)
                field("Guild Data", guildData)
                field("Uptime", "${d}d ${h}h ${m}m ${s}s")

                field("Users") {
                    buildString {
                        append("Total: ").appendln(users)
                        append("Online: ").appendln(online)
                        append("Offline: ").appendln(offline)
                        append("Inactive: ").appendln(inactive)
                    }
                }

                field("Others") {
                    buildString {
                        appendln("Creators: **[Avarel](https://github.com/Avarel)** and **[Xevryll](https://github.com/xevryll)**")
                        appendln("Contributor: **[Gatt](https://github.com/RealGatt)**")
                        appendln("Commands: **$commandSize**")
                        appendln("Library: Java **[JDA 3](https://github.com/DV8FromTheWorld/JDA)**")
                    }
                }
            }
        }.rest().queue()
    }
}
